fuzz: wallet, add target for Spend

Ayush170-Future opened the pull request:
This PR adds fuzz coverage for `wallet/spend`.

Motivation: Issue [27272](https://github.com/bitcoin/bitcoin/issues/27272#issue-1628327906)

This PR adds Fuzz coverage to the whole concept of Creating a New Transaction as well as other sections of the Spend file.  Because `CreateTransaction` is one of the most frequently used functions in the wallet codebase, merging this PR will significantly improve the wallet codebase's Fuzz testing!

I also used the `Singleton Class` concept for creating Wallet instances because it assures that only one instance of it is created during all Fuzz runs, which significantly boosts the file's `exec/sec`.

Ayush170-Future committed:
fuzz: add target for Spend

brunoerg commented on src/wallet/test/fuzz/spend.cpp:
```diff
@@ -0,0 +1,254 @@
+
+    void InitializeWallet() {
+        auto& node{g_setup->m_node};
+        wallet = CreateSyncedWallet(*node.chain, WITH_LOCK(Assert(node.chainman)->GetMutex(), return node.chainman->ActiveChain()), g_setup->coinbaseKey);
```
I think that `CreateSyncedWallet` will call `SetupDescriptorScriptPubKeyMans()` which internally will make a seed.
```cpp
void CWallet::SetupDescriptorScriptPubKeyMans()
{
    AssertLockHeld(cs_wallet);

    if (!IsWalletFlagSet(WALLET_FLAG_EXTERNAL_SIGNER)) {
        // Make a seed
        CKey seed_key;
        seed_key.MakeNewKey(true);
        CPubKey seed = seed_key.GetPubKey();
        assert(seed_key.VerifyPubKey(seed));

        // Get the extended key
        CExtKey master_key;
        master_key.SetSeed(seed_key);

        SetupDescriptorScriptPubKeyMans(master_key);
    }
```

Not sure, but wouldn't it be non-deterministic? A solution would be to have a fuzz version of `CreateSyncedWallet` which calls `SetupDescriptorScriptPubKeyMans` passing a key that we created using buf.

brunoerg commented on src/wallet/test/fuzz/spend.cpp:
```diff
@@ -0,0 +1,254 @@
+                                        /*nAmount=*/ConsumeMoney(fuzzed_data_provider), 
+                                        /*fSubtractFeeFromAmount=*/fuzzed_data_provider.ConsumeBool()});
+                }
+                // Equally likely to chose one of the three values.
```
s/chose/choose

brunoerg commented on src/wallet/test/fuzz/spend.cpp:
```diff
@@ -0,0 +1,254 @@
+            },
+            [&] {
+                CAmount fee;
+                int random_positions[3] = {-1, fuzzed_data_provider.ConsumeIntegralInRange<int>(0, mtx.vout.size()), fuzzed_data_provider.ConsumeIntegral<int>()};
```
I think you could keep same pattern as you had previously:
```cpp
std::vector<int> random_positions = {-1, fuzzed_data_provider.ConsumeIntegralInRange<int>(0, recipients_size), fuzzed_data_provider.ConsumeIntegral<int>()};
int change_pos = PickValue(fuzzed_data_provider, random_positions);
```

instead of 
```cpp
int random_positions[3] = {-1, fuzzed_data_provider.ConsumeIntegralInRange<int>(0, mtx.vout.size()), fuzzed_data_provider.ConsumeIntegral<int>()};
int change_pos = fuzzed_data_provider.PickValueInArray(random_positions);
```

brunoerg commented:
From lint:
```diff
+ test/lint/all-lint.py
This diff appears to have added new lines with trailing whitespace.
The following changes were suspected:

diff --git a/src/wallet/test/fuzz/spend.cpp b/src/wallet/test/fuzz/spend.cpp
@@ -0,0 +1,254 @@
+namespace wallet { 
+ * Singleton wallet class ensures that only one 
+ * instance of CWallet is created and reused as required. 
+ * of creating a new `Descriptor Wallet` each time and deletes 
+CCoinControl GetNewCoinControl(FuzzedDataProvider& fuzzed_data_provider, CWallet& wallet, std::vector<COutput>& coins) 
+    
+    // Because none of the UTXO will ever be used (marked as spent), mining this many should be sufficient. 
+                                        /*nAmount=*/ConsumeMoney(fuzzed_data_provider), 
+                const CTxOut tx_out{ConsumeMoney(fuzzed_data_provider), ConsumeScript(fuzzed_data_provider)}; 
+    
+                // Taking a random sub-sequence from available coins 
+}   
Success: no issues found in 269 source files
^---- failure generated from lint-whitespace.py
```

Ayush170-Future commented:
Thanks, @brunoerg for your reviews! In the recent push, I've addressed most of your reviews.

- `clang-format` to fix the lint's issue.
- Changed `PickValueFromArray` to the `PickValue` approach.
- Fixed one grammar mistake.

Your [comment](https://github.com/bitcoin/bitcoin/pull/28236#discussion_r1286425941) is really interesting, but I'm not sure whether the `non-determinism` in this case might pose a significant issue or not.
So, I'll wait for more people to give their thoughts on this and we can write a new `CreateSyncedWallet` specific for Fuzzing purposes if felt necessary.

Ayush170-Future commented:
The `fuzzer` logs are unclear, I'm not sure why it failed suddenly. Maybe some issue with the Fuzzer again?

Ayush170-Future commented:
- Force pushed rebasing on current master branch.

Ayush170-Future commented:
- Force-pushed updating the `CRecipient` wrt [28246](https://github.com/bitcoin/bitcoin/pull/28246).

All CI checks passed now 💪

dergoegge commented on src/wallet/test/fuzz/spend.cpp:
```diff
@@ -0,0 +1,259 @@
+        return instance;
+    }
+
+    CWallet& GetWallet()
```
```suggestion
    const CWallet& GetWallet()

```

It's otherwise hard to tell if the global wallet instance is mutated. Same for `GetCoins()`.

If the global instance is mutated, the target is likely to be non-deterministic which hurts fuzzer efficiency. The target should behave exactly the same when given the same input, but that might not be the case if global state is used and modified.

(The suggested change might not compile if `wallet` or `available_coins` are used as non-const/mutated)

Ayush170-Future replied:
Yeah, this should make it more deterministic. As you said, it's causing some compilation errors with other code. I'll force-push once I've fixed them.

Ayush170-Future replied:
The `CreateTransaction()` and `FundTransaction()` want the `CWallet` parameter to be non-const. I'm struggling to find a work around but I'm trying.

Just wanted to clear this, I think the `CWallet` instance inside these functions doesn't seem to be directly mutated or changed anytime. So, do we really want to make it to const. Am I understanding this correct?

Ayush170-Future replied:
I've thought about it, and I think a global Singleton might not be the smartest way to conserve resources. The `wallet` object definitely mutates during execution. I think I've to remove the Singleton pattern, but doing so will make the file very slow, as creating the `Descriptor Wallet` is really time-consuming.

dergoegge reviewed (changes requested):
Coverage looks decent: https://dergoegge.github.io/bitcoin-coverage/pr28236/fuzz.coverage/index.html

maflcko reviewed (commented):
Spend.cpp coverage is not too bad already: https://maflcko.github.io/b-c-cov/fuzz.coverage/src/wallet/spend.cpp.gcov.html

It may be good to clarify the new coverage a bit.

Ayush170-Future commented on src/wallet/test/fuzz/spend.cpp:
```diff
@@ -0,0 +1,259 @@
+    }
+
+    // Plays with the `CoinsResult`
+    CoinsResult wallet_coins;
+
+    LIMITED_WHILE(fuzzed_data_provider.ConsumeBool(), 50)
```
Yeah, I can create a separate PR for this immediately. Once it's ready, we can merge it, and I can continue working on this. Thanks!

murchandamus replied:
Agree wtih @maflcko, since the seeds are created per target, it would be better to have a separate target for `CoinsResult`.

Ayush170-Future commented:
> Spend.cpp coverage is not too bad already: https://maflcko.github.io/b-c-cov/fuzz.coverage/src/wallet/spend.cpp.gcov.html
> 
> It may be good to clarify the new coverage a bit.

I didn't understand. Are you suggesting that I reconsider what I want to increase coverage of in this file?

Also, I have a question about the blue lines in the coverage reports. Do they represent the line coverage of the code during fuzzing? For instance, if these functions are encountered in the process of running any fuzz file, the fuzz coverage will display them in blue. If I'm understanding this correctly, wouldn't it still make sense to have a separate fuzz file for `Creating Transaction` so that we can explore more test cases and code flows? These might not be tested if the coverage is coming from another fuzz file designed for a different purpose or codebase.

 https://maflcko.github.io/b-c-cov/fuzz.coverage/src/wallet/spend.cpp.gcov.html

murchandamus reviewed (commented):
Hey,

Just did a quick pass. What is the status of this PR?

brunoerg commented on src/wallet/test/fuzz/spend.cpp:
```diff
@@ -0,0 +1,259 @@
+        }
+    }
+
+    // Occasionally, some random `CCoutPoint` are also selected.
```
```suggestion
    // Occasionally, some random `COutPoint` are also selected.

```

Ayush170-Future commented:
> Are you still working on this?

Sorry I got a little occupied with some other things and would want to continue working on this. I've the approach in mind and will force push the update by this end of this month.

- Will create a new file for `CoinsResult ` and remove global Singleton from the `Spend` file.

fanquake commented:
> I've the approach in mind and will force push the update by the end of this month.

Moved to draft for now.
